deploy

```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/env bash
+[[ -n "$TRACE" ]] && set -x
+set -eo pipefail
+
+readonly SOURCE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+readonly APPS_DIR="${SOURCE_DIR}/apps"
+readonly DASH_ENTERPRISE_URL="${DASH_ENTERPRISE_URL:-https://dash-playground.plotly.host}"
+readonly CREATE_APP="${CREATE_APP:-false}"
+readonly ALLOW_DASHR="${ALLOW_DASHR:-false}"
+
+log-header() {
+  declare desc="Log header formatter";
+  echo "====> $*"
+}
+
+log-info() {
+  declare desc="Log info formatter";
+  echo "      $*"
+}
+
+log-warn() {
+  declare desc="Log warn formatter";
+  echo " ?    $*" 1>&2
+}
+
+log-fail() {
+  declare desc="Log fail formatter";
+  echo " !    $*" 1>&2
+  exit 1
+}
+
+log-exit() {
+  declare desc="Log exit formatter";
+  echo " !    $*" 1>&2
+  exit 0
+}
+
+fn-check-env() {
+  declare APP="$1"
+
+  if grep -Fxq "$APP" "$SOURCE_DIR/.deployignore"; then
+    log-exit "App $APP is in the .deployignore file, skipping deploy"
+  fi
+
+  if ! command -v "dds-client" >/dev/null; then
+    log-fail "dds-client binary not found"
+  fi
+
+  if [[ -z "$DASH_ENTERPRISE_API_KEY" ]]; then
+    log-fail "DASH_ENTERPRISE_API_KEY is not defined"
+  fi
+
+  if [[ -z "$DASH_ENTERPRISE_USERNAME" ]]; then
+    log-fail "DASH_ENTERPRISE_USERNAME is not defined"
+  fi
+
+  if [[ -z "$DASH_ENTERPRISE_URL" ]]; then
+    log-fail "DASH_ENTERPRISE_URL is not defined"
+  fi
+
+  if [[ -z "$CIRCLE_SHA1" ]]; then
+    log-fail "CIRCLE_SHA1 is not defined"
+  fi
+}
+
+main() {
+  declare APP="$1"
+  local app_dir="$APPS_DIR/$APP"
+  local remote_url="$DASH_ENTERPRISE_URL/GIT/$APP"
+  local is_dashr=false
+  local tmp_shallow_clone="$(mktemp -d -t apps-XXXXXXXXXX)"
+  local tmp_full_clone="$(mktemp -d -t apps-XXXXXXXXXX)"
+  local force_push=false
+  local push_code=false
+  local exit_code remote_sha
+  trap "rm -rf '$tmp_shallow_clone' '$tmp_full_clone' >/dev/null" RETURN INT TERM EXIT
+
+  fn-check-env "$APP"
+
+  if [[ -z "$APP" ]]; then
+    log-fail "No app name specified"
+  fi
+
+  if [[ ! -d "$app_dir" ]]; then
+    log-fail "Invalid app name: $APP"
+  fi
+
+  if [[ "$APP" =~ "dashr" ]]; then
+    is_dashr=true
+    if [[ "$ALLOW_DASHR" != "true" ]]; then
+      log-exit "App $APP is a dashr app, skipping deploy"
+    fi
+  fi
+
+  if [[ ! -d "$app_dir/assets" ]]; then
+    log-fail "App $APP is missing the assets directory"
+  fi
+
+  log-header "Deploying $APP"
+  log-info "dashr: ${is_dashr}"
+  if [[ "$is_dashr" == "false" ]]; then
+    local celery_version="$(cat "$app_dir/requirements.txt" | grep -E "^celery=")"
```

Comment:
These can also take the form `celery>={version}`!

Reply:
We actually don't use celery in any of these apps, but once I add this to other monorepos, I'll keep that in mind - and come back to update this one.